### PR TITLE
Add CRM_Core_Form::isFormInViewMode and CRM_Core_Form::isFormInEditMode

### PR DESCRIPTION
**Pull request by mlutfy**
Overview
----------------------------------------

Extensions implementing the buildForm hook often have to make sure that `$form->elementExists()` before applying a certain logic, otherwise QuickForm will throw a fatal error if the field is missing.

Sometimes we use it as a lazy way of checking the mode of the form, which we can get through `$form->getMode()` and then try to remember which of the `CRM_Core_Action` constants we have to check against.

I stumbled on `CRM_Core_Form::isFormInViewOrEditMode` and found it rather practical, but it is set as protected, so we cannot call it from a buildForm hook, which has `$form`, not `$this`.

Comments
----------------------------------------

The function was added in 2018-01 as part of 240b0e652adbcfc01ea0f97989ba5cc74d3c7b79, which is fitting. I stumbled on a similar bug in an extension just now, while trying to delete a profile field.

### CODE REVIEW

**Comment by mattwire** [comment]
@mlutfy Not blocking and happy to merge. But does it make sense to have `isInViewMode()` and `isInEditMode()` instead?

**Comment by mlutfy** [comment]
@mattwire Maybe it's lazyness on my part, but I mostly need a way to avoid running buildForm code during a deletion, and that function seems to work for most use-cases.

(Form View is still a normal form, just usually frozen, and BROWSE/BASIC/ADVANCED/PREVIEW are rarely used).

**Comment by eileenmcnaughton** [comment]
@mlutfy can we please do what @mattwire suggested

**Comment by mlutfy** [comment]
OK, pushed an update :)

**Comment by eileenmcnaughton** [comment]
As an aside the 'ready for review' tag should probably be added by someone other than the PR submitter. The fact the PR exists & doesn't have [WIP] in the header means the submitter thinks it's ready. The tag should indicate someone else does as well

**Comment by mattwire** [comment]
test this please

**Comment by mlutfy** [comment]
Makes sense - although it's not easy to see PRs where feedback was given, and the author responded. Then again, most authors can't add tags, so it makes sense to keep this tag for reviewers (and not self-tag).

**Comment by mattwire** [comment]
> it's not easy to see PRs where feedback was given, and the author responded

@mlutfy That's what I created the tag for and intended it be used for. I add it on my own PRs and others in exactly that situation.

**Comment by demeritcowboy** [comment]
Is the function contents mismatch going to be confusing? I would expect from the wording that if someone were to refactor

```
function isFormInViewOrEditMode() {
  return $form->isFormInViewMode() || $form->isFormInEditMode();
}
```
that it would be identical, but that wouldn't always be true here. Is there a reason for leaving out BASIC etc? You said they're not often used, but are they *never* used when this is called?

**Comment by eileenmcnaughton** [comment]
Yeah - the thing about 'ready to review' is that it can't be added by un-permissioned users - whereas pings & 'request review ' can it implies that these ones are ready & others are not. 

I don't personally factor it's existence or not into whether I look at a PR - since I haven't found PRs tagged with it as any more ready-to-review than others , but some people may do.

**Comment by eileenmcnaughton** [comment]
Perhaps we should automatically tag ALL PRs as 'ready for review' & then we can REMOVE the tag if it turns out not to be?

**Comment by eileenmcnaughton** [comment]
@mlutfy I've removed the MOP because there is feedback by @demeritcowboy - can you check that out  respond to it.

**Comment by eileenmcnaughton** [comment]
@mlutfy also - please self-merge this if you can agree it with @demeritcowboy  - I don't think we need further involvement by others as whatever you & @demeritcowboy agree seems fine in this context

**Comment by mlutfy** [comment]
@demeritcowboy Good point - I think those actions were added for "action links", and most of them are basically a type of "Edit" mode. I updated my patch.

File: CRM/Core/Form.php [hunk 1]
```diff
@@ -2576,17 +2576,40 @@ public function getCurrency($submittedValues = []) {
    * @return bool
    */
   protected function isFormInViewOrEditMode() {
+    return $form->isFormInViewMode() || $form->isFormInEditMode();
```
Comment (demeritcowboy) [inline CRM/Core/Form.php:2579]: Should be `$this`. Sorry my sample code was out of context.

Comment (mlutfy) [reply CRM/Core/Form.php]: yikes, my bad!

**Comment by demeritcowboy** [comment]
Thanks looks good. As per earlier comments "merge on pass".

**Comment by mlutfy** [comment]
Thanks! It's a really minor detail, but I often get bit by this in extensions. It'll make things cleaner!

**Pull request merged**